File: BST/C++/node.cpp
#include <algorithm>

class Node{
    public: int value;
    private: Node* left;
    private: Node* right;

    public: Node(int value) : value(value), left(nullptr), right(nullptr){}

    public: bool add(int value){
        if(value < this->value){
            if(this->left == nullptr){
                this->left = new Node(value);
                return true;
            }
            return this->left->add(value);
        }

        if(value > this->value){
            if(this->right == nullptr){
                this->right = new Node(value);
                return true;
            }
            return this->right->add(value);
        }

        return false;
    }

    public: bool contain(int value){
        if(value < this->value){
            if(this->left == nullptr) return false;
            return this->left->contain(value);
        }

        if(value > this->value){
            if(this->right == nullptr) return false;
            return this->right->contain(value);
        }

        return true;
    }

    public: int length(){
        int left_len = 0;
        if(this->left != nullptr){
            left_len = this->left->length();
        }

        int right_len = 0;
        if(this->right != nullptr){
            right_len = this->right->length();
        }
        return left_len + right_len + 1;
    }

    public: int height(){
        int left_height = 0;
        if(this->left != nullptr){
            left_height = this->left->height();
        }

        int right_height = 0;
        if(this->right != nullptr){
            right_height = this->right->height();
        }

        return 1 + std::max(left_height, right_height);
    }


};

File: BST/golang/node.go
package main

type Node struct {
	value int
	left *Node
	right *Node
}

func (self *Node) Add(value int) bool {
	if value < self.value {
		if self.left == nil {
			self.left = &Node{value: value, left: nil, right: nil}
			return true
		}
		return self.left.Add(value) 
	}

	if value > self.value {
		if self.right == nil {
			self.right = &Node{value: value, left:nil, right:nil}
			return true
		}
		return self.right.Add(value)
	}
	return false
}

func (self *Node) Contain(value int) bool {
	if value < self.value {
		if self.left == nil {
			return false
		}
		return self.left.Contain(value)
	}

	if value > self.value {
		if self.right == nil {
			return false
		}
		return self.right.Contain(value)
	}
	return true
}

func (self *Node) Length() int {
	a := 0
	if self.left != nil {
		a = self.left.Length()
	}
	b := 0
	if self.right != nil {
		b = self.right.Length()
	}
	return a + b + 1
}

func (self *Node) Height() int {
	a := 1
	if self.left != nil {
		a = self.left.Height() + 1
	}
	b := 1
	if self.right != nil {
		b = self.right.Height() + 1
	}
	if a > b{
		return a
	}
	return b
}